controller/src/main/java/org/jboss/as/controller/AbstractWriteAttributeHandler.java

```diff
@@ -87,15 +88,14 @@ public void execute(OperationContext context, ModelNode operation) throws Operat
         final AttributeDefinition attributeDefinition = getAttributeDefinition(attributeName);
         final ModelNode defaultValue;
         if (attributeDefinition != null) {
-            defaultValue = attributeDefinition.getDefaultValue();
             final ModelNode syntheticOp = new ModelNode();
+            defaultValue = attributeDefinition.getDefaultValue();
             syntheticOp.get(attributeName).set(newValue);
             attributeDefinition.validateAndSet(syntheticOp, submodel);
             newValue = submodel.get(attributeName);
             recordCapabilitiesAndRequirements(context, attributeDefinition, newValue, currentValue);
         } else {
-            defaultValue = null;
-            submodel.get(attributeName).set(newValue);
+            throw new OperationFailedException(ControllerLogger.MGMT_OP_LOGGER.nullNotAllowed(attributeName));
```

Comment:
Thank you, this one works too (just without the new parameter). Will commit this change together with next iteration of problem above

Reply:
This should throw an OFE. The class javadoc for OFE describes what it is for:

```
 * Exception indicating an operation has failed due to a client mistake (e.g. an operation with
 * invalid parameters was invoked.) Should not be used to report server failures.
 * 
```

It's not obvious, but this 'else' indicates a server failure. Really it indicates a coding failure. L88 should not return null. If it did return null either an ISE should be thrown or an 'assert' statement should fail.

@yersan This kind of use of OFE is something to watch out for. It's very, very common; people see that OFE is listed in the method signature so if a failure occurs that's what gets thrown. But that will result in the server mishandling the fault, because it treats the OFE as a client mistake not a server failure. RuntimeExceptions should be used for server failures.

Reply:
@petrberan I think what needs to be done is the javadoc of getAttributeDefinition should be changed so the @return doc says it will not return {@code null} and then add a '@throws IllegalArgumentException if no AttributeDefinition with name {@code attributeName} exists". Then change the impl to match. (Note that the 'attributeDefinitions' field cannot be null so the null check can be removed.)

The if/else were discussing here can then be removed, with just the 'if' body retained.

ReloadRequiredWriteAttributeHandler.convertToType can be modified to eliminate its null check on the return value. All other uses I see of that method in WF Core and in full WildFly all assume it doesn't return null; i.e. they would NPE if it did.

Reply:
@bstansberry Thank you for your input. So technically speaking, this solution is "correct" (but throws different exception at different place)?

Does that also mean that EnhancedSyntaxTestCase tests (such as performCompositeListWrite#755) should fail in runtime, since there's no attributeDefinition for enhanced syntax attribute?

Reply:
Yes, it's basically correct, but different exception at different place.

I may have missed something but I don't believe what I'm recommending here should break a proper test of enhanced syntax.  If something does break perhaps the test is flawed or the way we handle enhanced syntax is.

Enhanced syntax works by code that analyzes the user-provided attribute name and working out what the real attribute name is.  The AbstractWriteAttributeHandler.getAttributeDefinition method should be provided the real attribute name. It is not a requirement that the param value directly reflect what a user provided; this method is internal.

Reply:
EnhancedSyntaxTestCase fails several times because in WriteAttributeHandler#181 there's created a clone of runtime operation, which is resolved without EnhancedSyntax#186, but then the original operation is executed with EnhancedSyntax#197 again - effectively it tries to getAttributeDefinition on "runtime-list-attribute[-1]", which does not exist, only "runtime-list-attribute" does

Reply:
@bstansberry Notice that for a runtime attribute expressed in the following form:

`/resource:write-attribute(name=runtime-attribute[-1], value=value2)`

The `attributeName` at AbstractWriteAttributeHandler:87 is "runtime-attribute[-1]" because the attribute syntax resolution is deferred for the Runtime stage on WriteAttributeHandler:186.

I think we should handle the ISE error on `WriteAttributeHandler#doExecuteInternal` method instead of `AbstractWriteAttributeHandler:97` to take into account the model and runtime stages.

I am not sure if we are allowed to fail with an exception on a runtime stage, however, since this should be considered a coding error, maybe we are allowed to. If we are not allowed to throw the ISE at runtime, we could try to check this specific case (runtime attribute expressed as runtime-attribute[..] syntaxis) at the model stage and convert `runtime-attribute[-1]` into `runtime-attribute` before invoking AbstractWriteAttributeHandler:87

Reply:
Sorry, @yersan and @petrbaran, I realize I commented on a particular detail point (the exception type) and then kept saying more without properly digesting all the other discussion that had gone on, where you guys were dealing with enhanced syntax. :(

So, I'm not sure where to go here, other than I'll say this:

1)WriteAttributeHandler.execute in code up to L95 figures out the base attribute name and fails if it is not a registered attribute. So that's where we check for a client mistake in the base name. And then in the enhanced syntax case, later getEnhancedSyntaxResolvedOperation gets called which calls EnhancedSyntaxSupport.updateWithEnhancedSyntax, which in turns validates the rest of the enhanced syntax, i.e. checks for client mistakes there. I believe those two places handles cases of client mistakes related to the operation's 'name' value, so that's the only code that should throw OFE due to that kind of problem.

2) I did not expect any existing test to need to be changed for this JIRA. This was not meant to result in any behavior change. It's about eliminating old, should-be-unused code paths, and cleaning up any odd stuff we find when doing that. A perfect, but unexpected example apparently being how enhanced syntax works. It's possible some old bad test might need a tweak for some reason but we shouldn't be changing tests because of a belief that this JIRA is meant to change behavior.

3) But, this WriteAttributeHandler.doExecuteInternal stuff around runtime attributes and resolvedOperation and etc seems broken. L197 should not be calling write-handler.execute with an op whose 'name' is an expression.  Also the before-this-PR AbstractWriteAttributeHandler behavior where it reads a model from the Resource and then adds (via a get(attributeName) call) a 'runtime-attribute[-1]' child node is definitely wrong. It's basically just polluting that model with a garbage child node.

Perhaps the fix for that would be to do this replacing the current WriteAttributeHandler L182:

```
// We have to invoke on the AttributeAccess' write handler in Stage.MODEL but we don't know yet what the value
// we want to write is. We can't know that until Stage.RUNTIME.
// So, for Stage.MODEL we are going to pass to the AttributeAccess' write handler an operation that amounts to 
// writing the same value as exists now. The Stage.MODEL handling for a Storage.RUNTIME attribute is basically 
// meaningless, so passing in an operation whose value amounts to a no-op should be ok (although kludgy).
// But before calling that handler we will register a Stage.RUNTIME step that will change the 'operation' object to the 
// properly resolved operation value. The step we add will execute *before* any Stage.RUNTIME step the write handler 
// adds, so the one the write handler adds will see the operation state that our step will set up. 
final ModelNode resolvedOperation = operation.clone();
resolvedOperation.get(NAME).set(attributeName); // replace the current expression with the base attribute name
resolvedOperation.get(VALUE).set(currentValue); // just pass in the existing value
```

This is hacky but it at least isolates this strangeness to WriteAttributeHandler (and explains a bit what is going on.)

Reply:
Thank you both for your help, I will iterate on this problem tomorrow. Just a quick note after today: 

I added a syntax and requiresRuntime (AbstractWriteAttributeHandler#254) check into execute method, which will extract attributeName in case of both arguments are true. This "solution" ensures that all EnhancedSyntaxTestCase test will not fail and the original logic remains unchanged. 

Had to try said solution just to see if any other tests will be broken by this issue - it made two tests in ModelControllerMBeanTestCase fail. For some unknown reasons, there's a custom handler at ModelControllerResourceDefinition#156 that did not pass attribute definition of an attribute. Changed this for a default one, tests pass.

I will commit those changes shortly.

After those "fixes", all tests run without failure and I'm able to build the project without any issues.

Reply:
Tried proposed solution, it seems it doesn't work. For runtime in Stage.MODEL the operation with enhanced syntax is executed, causing tests to fail. I tried tweaking this solution a little bit to no good, resulting in only more and/or different errors in those tests.

If those tests are correct, my code seems to work according to them, so I just polished it a little bit more.